pyiceberg/avro/resolver.py

```diff
@@ -233,7 +255,107 @@ def skip(self, decoder: BinaryDecoder) -> None:
         pass
 
 
-class SchemaResolver(PrimitiveWithPartnerVisitor[IcebergType, Reader]):
+class WriteSchemaResolver(PrimitiveWithPartnerVisitor[IcebergType, Writer]):
+    def schema(self, schema: Schema, expected_schema: Optional[IcebergType], result: Writer) -> Writer:
+        return result
+
+    def struct(self, struct: StructType, provided_struct: Optional[IcebergType], field_writers: List[Writer]) -> Writer:
+        if not isinstance(provided_struct, StructType):
+            raise ResolveError(f"File/write schema are not aligned for struct, got {provided_struct}")
+
+        provided_struct_positions: Dict[int, int] = {field.field_id: pos for pos, field in enumerate(provided_struct.fields)}
```

Comment:
Is this backward? `provided_struct` is the file schema The `StructWriter` uses `pos` to look up the data in the incoming struct, so it should be the position in `struct`, the data's schema.

Reply:
Okay, it was backward. I thought that the file schema was the partner schema, but it is actually the schema being traversed, which is shown in the test case. I think it does need to be clear which schema is which.

Reply:
Yes, this was a bit messy because I went back and forth a couple of times. Just cleaned it up.